Various stuff done, see summary created by CodeRabbit as it is quite good

Laurentiu-Andronache opened the pull request:
- Implemented a fallback mechanism to extract navigation links from the entire page when traditional navigation structures are absent, such as in Mintlify sites.
- Enhanced link filtering to ensure only meaningful documentation links are included, avoiding duplicates and irrelevant URLs.

This update improves the reliability of the navigation extraction process.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Modular, extensible navigation detection for multiple docs platforms and depth-aware ToC with section headers.

* **Bug Fixes**
  * Improved URL normalization, deduplication, ordering and depth handling for consistent traversal and ToC generation.

* **Chores**
  * CLI simplified by removing legacy recursive flag; downloader now accepts explicit markdown mode.

* **Tests**
  * Added a test script to automate bulk documentation downloads.

* **Documentation**
  * README revamped for multi-platform usage, extractors guide, testing and updated usage examples.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Laurentiu-Andronache committed:
Add fallback navigation extraction for GitbookDownloader

- Implemented a fallback mechanism to extract navigation links from the entire page when traditional navigation structures are absent, such as in Mintlify sites.
- Enhanced link filtering to ensure only meaningful documentation links are included, avoiding duplicates and irrelevant URLs.

This update improves the reliability of the navigation extraction process.

Amal-David commented on gitbook_downloader.py:
```diff
@@ -348,6 +348,36 @@ async def _extract_nav_links(self, content):
+                        if (full_url not in processed_urls and
+                            full_url != self.base_url.rstrip("/") and
+                            full_url.startswith(self.base_url) and
+                            not any(skip in full_url for skip in ["#", "mailto:", "tel:", ".pdf", ".jpg", ".png", ".gif", ".svg", "api-docs"])):
```
@Laurentiu-Andronache can you resolve this and push?

Laurentiu-Andronache committed:
Enhance URL handling in GitbookDownloader to skip fragment-only links

- Updated duplicate URL detection to ignore fragments, ensuring that only unique base URLs are processed.
- Improved filtering to skip fragment-only links and links that only differ by fragments from the base URL, enhancing the accuracy of navigation link extraction.

Laurentiu-Andronache committed:
Enhance content extraction in GitbookDownloader for improved reliability

- Updated main content extraction logic to prioritize specific HTML containers, enhancing compatibility with various site structures.
- Added functionality to remove common sidebar elements and permalink anchor links, streamlining the final output.
- Improved table of contents generation by including h2 headings for single-page documents, enriching navigation for users.

Laurentiu-Andronache committed:
Refactor GitbookDownloader to simplify initialization and enhance link handling

- Removed the recursive option from the GitbookDownloader constructor, streamlining the initialization process.
- Updated link extraction logic to always search for sub-navigation links, improving compatibility with JavaScript-rendered navigation structures.
- Enhanced duplicate URL detection by normalizing trailing slashes, ensuring more accurate processing of links.

Laurentiu-Andronache committed:
Add test script for GitbookDownloader to verify functionality

- Introduced a new test script that downloads multiple documentation sites to ensure the downloader operates correctly.
- The script creates a new test folder for each run to preserve previous results, enhancing testing organization.
- Utilizes subprocess to execute the downloader for each specified documentation URL, improving automated testing capabilities.

Laurentiu-Andronache committed:
Enhance GitbookDownloader navigation extraction and content processing

- Introduced support for hierarchical navigation links, allowing for better organization of pages with depth tracking.
- Added handling for section headers in navigation, improving the representation of content structure.
- Updated the extraction logic to differentiate between sites with global navigation (e.g., Mintlify) and traditional structures, optimizing the download process.
- Enhanced markdown generation to include indentation based on page depth, improving readability of the generated content.
- Improved duplicate URL handling by ensuring unique entries while preserving order, enhancing the reliability of navigation extraction.

Laurentiu-Andronache commented:
I fixed what you said. I also created for you a test.py to be able to test faster four docs that I've found harder to crawl. We should definitely add a few more URLs in that list for the tests to be more comprehensive.

Laurentiu-Andronache committed:
Refactor navigation extraction using Strategy pattern

- Add normalize_url() and should_skip_url() utilities to deduplicate URL handling
- Create NavExtractor ABC defining can_handle() and extract() interface
- Implement MintlifyExtractor for docs with id="navigation-items"
- Implement GitBookExtractor for traditional nav/aside structures
- Implement FallbackExtractor for sites without standard navigation
- Simplify _extract_nav_links from 115 to 25 lines using extractor chain
- Remove unused imports: requests, json, Comment, os, ThreadPoolExecutor, partial

Laurentiu-Andronache committed:
Add VocsExtractor for hierarchical navigation in Vocs documentation sites

- Add VocsExtractor class to handle Vocs docs (e.g., metalex-docs.vercel.app)
  with proper hierarchy extraction from nested sidebar sections
- Update depth for pages when found at more accurate depth in nav extraction
- Add URL-based sorting for non-global-nav sites to group related pages
- Bootstrap into collapsed sections by also extracting content links
- Preserve index-based sorting for Mintlify sites (global nav)

Laurentiu-Andronache committed:
Update README to document plugin-based extractor architecture

- Rename to "Documentation Downloader" reflecting multi-platform support
- Add Supported Platforms section with extractor table
- Document hierarchical navigation, content extraction, rate limiting features
- Add Testing section for test.py usage
- Add "Adding New Extractors" guide with example code
- Add Architecture diagram showing extractor hierarchy
- Update Technical Details with current dependencies
